actionview/lib/action_view/unbound_template.rb

```diff
@@ -14,10 +14,26 @@ def initialize(source, identifier, details:, virtual_path:)
       @virtual_path = virtual_path
 
       @templates = Concurrent::Map.new(initial_capacity: 2)
+      @write_lock = Mutex.new
+      @templates = {}
     end
 
     def bind_locals(locals)
-      @templates[locals] ||= build_template(locals)
+      if template = @templates[locals]
+        template
+      else
+        @write_lock.synchronize do
+          normalized_locals = normalize_locals(locals)
+
+          # We need ||=, both to dedup on the normalized locals and to check
+          # while holding the lock.
+          @templates[normalized_locals] ||= build_template(normalized_locals)
+
+          # This may have already been assigned, but we've already de-dup'd so
+          # reassignment is fine.
+          @templates[locals.dup] = @templates[normalized_locals]
```

Comment:
I stared at this for a little too long. Can you confirm I am understanding this correctly?

We want to build the template exactly once for a given `normalized_locals`, so while holding the lock we build it if it doesn't already exist under the `normalized_locals` key. Then we also store it with the non-normalized locals so if we call `bind_locals` again with those same locals we don't need to normalize them again.

Reply:
That's correct! We want to avoid (what was previously) the requirement of normalizing keys (and duplicating the array) for every lookup.